networkx/algorithms/isomorphism/vf2pp_helpers/candidates.py

```diff
@@ -0,0 +1,61 @@
+def _find_candidates(u, graph_params, state_params):
+    """Given node u of G1, finds the candidates of u from G2.
+
+    Parameters
+    ----------
+    u: Graph node
+        The node from G1 for which to find the candidates from G2.
+
+    graph_params: namedtuple
+        Contains all the Graph-related parameters:
+
+        G1,G2: NetworkX Graph or MultiGraph instances.
+            The two graphs to check for isomorphism or monomorphism
+
+        G1_labels,G2_labels: dict
+            The label of every node in G1 and G2 respectively
+
+    state_params: namedtuple
+        Contains all the State-related parameters:
+
+        mapping: dict
+            The mapping as extended so far. Maps nodes of G1 to nodes of G2
+
+        reverse_mapping: dict
+            The reverse mapping as extended so far. Maps nodes from G2 to nodes of G1. It's basically "mapping" reversed
+
+        T1, T2: set
+            Ti contains uncovered neighbors of covered nodes from Gi, i.e. nodes that are not in the mapping, but are
+            neighbors of nodes that are.
+
+        T1_out, T2_out: set
+            Ti_out contains all the nodes from Gi, that are neither in the mapping nor in Ti
+
+    Returns
+    -------
+    candidates: set
+        The nodes from G2 which are candidates for u.
+    """
+    G1, G2, G1_labels, _, _, nodes_of_G2Labels, G2_nodes_of_degree = graph_params
+    mapping, reverse_mapping, _, _, _, _ = state_params
+
+    covered_neighbors = [nbr for nbr in G1[u] if nbr in mapping]
+    if not covered_neighbors:
+        return {
+            node
+            for node in G2.nodes()
+            if node not in reverse_mapping
+            and all(nbr2 not in reverse_mapping for nbr2 in G2[node])
```

Comment:
I am trying to figure out this line again: `and all(nbr2 not in reverse_mapping for nbr2 in G2[node])`
This is the case when no G1 neighbors of u are in the mapping (no covered_neighbors).
That means any potential candidates from G2 must not have any neighbors in the reverse_mapping.
Hence this line ensures that `node` does not have any neighbors in the reverse_mapping.

My question is:  aren't the nodes with no neighbors in the reverse_mapping exactly the nodes in `T2_out`?
If so, perhaps this line should be: `if node in state.T2_out`.

Also, I don't think we have to look at all the `G2.nodes()`.  Can we make this loop over a smaller set of nodes? Maybe we could only over the nodes with the right degree and label instead of all the nodes...
`for node in G2_nodes_of_degree[G1.degree[u]] & nodes_of_G2Labels[G1_labels[u]]`